Fix for issue #7645: Do not preserve 'cw' and 'ccw' attributes in PlanarEmbedding.to_undirected()

Schwarf opened the pull request:
I propose this fix for [issue #7645](https://github.com/networkx/networkx/issues/7645). The changes ensure that the 'cw' and 'ccw' attributes are not preserved for both graphs and multi-graphs in the result of PlanarEmbedding.to_undirected().

I have included tests to verify that the issue is resolved.

I would be grateful for any feedback or suggestions to improve this PR. I am more than happy to incorporate any necessary changes based on the review. Thank you for your time and guidance!

Fixes #7645

dschult commented:
Thank you for putting this together @Schwarf 

These changes are made to the base class methods. But they should only apply for the PlanarEmbedding subclass. That is, someone who happens to create edge attributes named 'cw' will lose those attributes when they haven't ever used the planarity code. Can you move these changes to methods in the PlanarEmbedding class?

Schwarf commented:
Hi,
thank you for your feedback. I will have a look.

Schwarf commented:
My update has now added an overload of the `to_undirected()` method to the class `PlanarEmbedding`. I have removed the parameter ```reciprocal``` and corresponding functionality, because in my understanding its not relevant for planarity. Am I correct? 
For the `as_view `parameter I am actually not sure if its useful or not. I left the parameter but did not write a test (yet), until ist confirmed that this functionality is wanted in class `PlanarEmbedding`.

Thank you for your support. Looking forward to your feedback.

rossbar commented on networkx/algorithms/tests/test_planarity.py:
```diff
@@ -291,6 +291,27 @@ def test_edge_removal_from_planar_embedding(self):
         P.remove_edges_from(((0, 3), (1, 3)))
         self.check_graph(P, is_planar=True)
 
+    def test_graph_planar_embedding_to_undirected(self):
+        G = nx.Graph(((0, 1), (1, 2), (2, 3), (3, 0), (0, 2)))
+        is_planar, P = nx.check_planarity(G)
+        assert is_planar is True
+        U = P.to_undirected()
+        assert U[0][1] == {}
+        assert U[0][2] == {}
+        assert U[0][3] == {}
+        assert U[2][3] == {}
+        assert U[1][2] == {}
+
+    def test_multigraph_planar_embedding_to_undirected(self):
+        G = nx.MultiGraph([(1, 2), (1, 2), (1, 2), (1, 2), (2, 3), (3, 1)])
+        is_planar, P = nx.check_planarity(G)
+        assert is_planar is True
+
+        U = P.to_undirected()
+        assert U[1][2] == {}
+        assert U[2][3] == {}
+        assert U[3][1] == {}
+
```
A few minor suggestions to combine/simplify the tests - not a blocker!

```suggestion
    @pytest.mark.parametrize("graph_type", (nx.Graph, nx.MultiGraph))
    def test_graph_planar_embedding_to_undirected(self, graph_type):
        G = graph_type([(0, 1), (0, 1), (1, 2), (2, 3), (3, 0), (0, 2)])
        is_planar, P = nx.check_planarity(G)
        assert is_planar
        U = P.to_undirected()
        assert isinstance(U, nx.Graph)
        assert all((d == {} for _, _, d in U.edges(data=True)))
```

rossbar reviewed (commented):
> My update has now added an overload of the to_undirected() method to the class PlanarEmbedding. I have removed the parameter reciprocal and corresponding functionality, because in my understanding its not relevant for planarity. Am I correct?
For the as_view parameter I am actually not sure if its useful or not. I left the parameter but did not write a test (yet), until ist confirmed that this functionality is wanted in class PlanarEmbedding.

My immediate reaction is that both `as_view` and `reciprocal` should be supported, if for no other reason than to maintain consistency with the API of the base class.

Re: `as_view` - AFAIK there's no (builtin) way to filter edge *attributes* in graph views, so an `as_view=True` will result in a graph view that still has the `cw` and `ccw` attributes on the edges. My current opinion is that this shouldn't be a blocker for now.

Re: `reciprocal` - I think it's a worthwhile feature to keep, even if it's not directly relevant for any sort of planarity analyses.

dschult commented on networkx/algorithms/planarity.py:
```diff
@@ -1400,3 +1401,63 @@ def copy(self, as_view=False):
+
+    def to_undirected(self, reciprocal=False, as_view=False):
+        """
+        Returns an undirected representation of the planar embedding for directed graphs.
```
My understanding is that the PlanarEmbedding class only works for undirected graphs. Even though it is a subclass of `DiGraph`, it is representing undirected graphs. So I think this wording should change to something like "non-embedded undirected graph":
```suggestion
        Returns a non-embedding undirected representation of the graph.
```

dschult commented on networkx/algorithms/planarity.py:
```diff
@@ -1400,3 +1401,63 @@ def copy(self, as_view=False):
+        """
+        Returns an undirected representation of the planar embedding for directed graphs.
+
+        This method extends the functionality of `DiGraph.to_undirected()` by ensuring that
+        the `"cw"` and `"ccw"` attributes of the half-edges are excluded from the edge data
+        in the resulting undirected graph. This is specific to the requirements of planar embeddings.
```
I think referring to DiGraph here is confusing. Can we reduce the description to just the edge attributes? Something like:
```suggestion
        While creating the undirected graph, all edge attributes are retained except 
        the `"cw"` and `"ccw"` attributes which are removed from the edge data.
        Those attributes are specific to the requirements of planar embeddings.
```

dschult commented on networkx/algorithms/planarity.py:
```diff
@@ -1400,3 +1401,63 @@ def copy(self, as_view=False):
+        reciprocal : bool (optional)
+          If True only keep edges that appear in both directions in the original digraph.
+        as_view : bool (optional, default=False)
+            If True, return an undirected view of the original directed graph.
+            This does not create a new graph but provides a view into the original graph.
```
```suggestion
            If True, return an undirected view of the PlanarEmbedding.
            This does not create a new graph but provides a view into the PlanarEmbedding.
```

dschult commented on networkx/algorithms/planarity.py:
```diff
@@ -1400,3 +1401,63 @@ def copy(self, as_view=False):
+
+        Returns
+        -------
+        G : Graph
+            An undirected graph with the same name and nodes as the original directed graph.
+            Edges are included with their data, except for the `"cw"` and `"ccw"` attributes,
+            which are omitted.
```
```suggestion
        G : Graph
            An undirected graph with the same name and nodes as the PlanarEmbedding.
            Edges are included with their data, except for the `"cw"` and `"ccw"` attributes,
            which are omitted.
```

dschult reviewed (commented):
Thanks for putting it into the PlanarEmbedding class. 
The code looks fine. I probably would have made the `as_view` and `reciprocal` raise exceptions if used, but this is fine too. (the concept of views of planar embeddings seems strange to me; and all valid embeddings should have half-edges in both directions for every edge so reciprocal would only be needed for invalid embeddings.)

My suggestions are about the docs -- mostly about reference to "the original graph" being shifted to "the PlanarEmbedding" and related shifts away from the notation that the planar graph is some sort of directed graph when it is actually a representation of an undirected graph that happens to use a DiGraph super-class.  You can change these suggestions by using local changes/commits. The github "add suggestion" feature makes it easy for me to suggest, but not as easy for you to adapt that to what you want. So use whatever you need to.

Schwarf commented:
Thank you for the detailed feedback! I appreciate your points, especially regarding the wording in the documentation. You’re absolutely correct that my description of the method was misleading, and I will make sure to update it to reflect the correct terminology and intent, specifically shifting references to "the original graph" to "the PlanarEmbedding" and clarifying the representation of undirected graphs.

I also think your suggestion to raise exceptions for as_view and reciprocal is an excellent idea. Providing immediate feedback to users about the unusual nature of the PlanarEmbedding graph is a great way to improve usability and prevent potential confusion. I’ll incorporate this change into the method as well.

dschult commented on networkx/algorithms/planarity.py:
```diff
@@ -1400,3 +1401,65 @@ def copy(self, as_view=False):
+         ----------
+         reciprocal : bool (optional)
+             Not supported for PlanarEmbedding. This parameter raises an exception
+             if used. All valid embeddings include half-edges by definition, making this
+             parameter unnecessary."
```
```suggestion
             if used. All valid embeddings include reciprocal half-edges by definition,
             making this parameter unnecessary.
```

dschult commented on networkx/algorithms/planarity.py:
```diff
@@ -1400,3 +1401,65 @@ def copy(self, as_view=False):
+        if reciprocal:
+            raise ValueError(
+                "'reciprocal=True' is not supported for PlanarEmbedding. "
+                "All valid embeddings include half-edges by definition, making this "
+                "parameter unnecessary."
```
```suggestion
                "All valid embeddings include reciprocal half-edges by definition, "
                "making this parameter unnecessary."
```

dschult commented on networkx/algorithms/planarity.py:
```diff
@@ -1400,3 +1401,65 @@ def copy(self, as_view=False):
+            raise ValueError("'as_view=True' is not supported for PlanarEmbedding.")
+
+        graph_class = self.to_undirected_class()
+        if as_view is True:
+            return nx.graphviews.generic_graph_view(self, graph_class)
```
```suggestion
```

dschult reviewed (commented):
Thanks for these changes. I only have minor tweaks here.

dschult reviewed (approved):
Looks good to me!
Thanks!!

Schwarf commented:
Hi @dschult @rossbar ,

I hope you're doing well. I wanted to clarify the process regarding issue #7645 linked to this PR. Should I close the issue manually once the PR is merged, or is it handled automatically/by the maintainers?

Thanks for your guidance, and Happy Holidays.

Best regards,
Andreas

dschult commented:
The automated close occurs if you put special text into the PR posting that connects the PR to the issue, or if someone has connected this PR to the issue in the Development section on the right sidebar of this page.  Those are the two ways I know of.

To use text to connect the PR to the issue, put `Fixes #7645` in the original post. I have edited the initial post to include that now.  I think it might also work by putting it into the commit description. And there are other syntaxes like `Closes #7645` that work too.  
[See details in the github docs](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue)

rossbar reviewed (approved):
LGTM,thanks @Schwarf and @dschult !
